Add a simple calendar implementation that supports streak highlighting.

greinard opened the pull request:
## Overview

This branch adds a simple calendar implementation that supports streak highlighting.

`SimpleCalendar` wraps the existing `Calendar` component, providing a default and state-based day rendering implementation.

It works by using a state configuration mapping (which maps a given state to a style and streak highlighting settings) as well as a function to compute the state for a given day.

![image](https://github.com/CareEvolution/MyDataHelpsUI/assets/5408603/652f2881-d091-4d67-8641-aa5d96c7ea73)

## Security

REMINDER: All file contents are public.

- [x] I have ensured no secure credentials or sensitive information remain in code, metadata, comments, etc.
  - [x] Please verify that you double checked that .storybook/preview.js does not contain your participant access key details. 
  - [x] There are no temporary testing changes committed such as API base URLs, access tokens, print/log statements, etc.
- [x] These changes do not introduce any security risks, or any such risks have been properly mitigated.

Describe briefly what security risks you considered, why they don't apply, or how they've been mitigated.

- No new security risks have been introduced.  This update just provides a default implementation for calendar day rendering that allows for some state based styling customization.

## Checklist

### Testing
- [ ] MyDataHelps iOS app
- [ ] MyDataHelps Android app
- [ ] [MyDataHelps website](mydatahelps.org)

### Documentation
- [x] I have added relevant Storybook updates to this PR as well.

chrisnowak commented on src/components/presentational/SimpleCalendar/SimpleCalendar.css:
```diff
@@ -0,0 +1,35 @@
+.mdhui-simple-calendar {
+    background: #fff;
```
Would remove background; it is up to the card or container what the background is typically...and if we really need a contrast we would want to use the css background vars anyways so it works in dark mode

greinard replied:
This is gone now as part of the PR revisions.

chrisnowak commented on src/components/presentational/SimpleCalendar/SimpleCalendar.tsx:
```diff
@@ -0,0 +1,68 @@
+    innerRef?: React.Ref<HTMLDivElement>;
+}
+
+export default function (props: SimpleCalendarProps) {
```
I would go with "StreakCalendar" I think.  I think the novelty here is in streaks...you could theoretically use this without streaks...but it would not be massive benefit necessarily...the reason that we would build this as opposed to some sort of standalone "CalendarDay" component is to wrap the logic of streaks

chrisnowak commented on src/components/presentational/SimpleCalendar/SimpleCalendar.tsx:
```diff
@@ -0,0 +1,68 @@
+            }
+        }
+
+        return <div className={dayClasses.join(' ')} style={dayStyle} onClick={() => props.onDayClicked(date)}>
```
This could be a separate component for "CalendarDay" that takes props for "streakLeft", "streakRight", "day",  "backgroundColor", "streakColor"

greinard commented:
@chrisnowak - I ended up reworking this a bit after seeing your suggestions.  I created a new `CalendarDay` component, but I went a step further and just moved all of the streaking logic into that.

We were going to need project specific calendar impls to define the state/styling config anyways, so it was already a bit awkward to have an additional "SomeKindOfCalendar" component in the hierarchy.  With this update, the project specific calendar impls can just define their state/styling configuration and state computing function and then use the new `CalendarDay` component to render the days, like so:

```
const renderDay = (year: number, month: number, day?: number): React.JSX.Element => {
	return <CalendarDay
		year={year}
		month={month}
		day={day}
		stateConfiguration={stateConfiguration}
		computeStateForDay={computeStateForDay}
		onClick={onDayClicked}
	/>;
};
```

In additional to some standard stories to show how the `CalendarDay` component renders on its own, I have also added a story that shows how to use it with the existing `Calendar` component.

reneefromhold commented on src/components/presentational/CalendarDay/CalendarDay.stories.tsx:
```diff
@@ -0,0 +1,254 @@
+    </div>
+</Layout>;
+
+const now = parseISO('2023-11-29T21:41:43.678Z');
```
nit: use today

greinard replied:
I renamed it to `testDate`.  It doesn't really represent "now" or "today".  It used to be `new Date()`, which worked for "now", but I had to set it to a fixed date.  Otherwise, depending on which day of the week/month it was, the story renderings would be different.  Now they are static.

reneefromhold commented on src/components/presentational/CalendarDay/CalendarDay.tsx:
```diff
@@ -0,0 +1,62 @@
+    let dayStyle: CSSProperties | undefined;
+    if (currentDayStateConfiguration?.streak) {
+        if (currentDayState === previousDayState && currentDayState === nextDayState) {
+            if (date.getDate() > 1 && date.getDay() > 0 && date.getDay() < 6 && date.getDate() < getLastDayOfMonth(date)) {
```
nit, this would read easier if there were variables for isFirstDayOfMonth, isLastDayOfWeek etc instead of date.getDate() > 1, date.getDay() < 6

greinard replied:
Good call.  I reworked the logic a bit to simplify it and pulled out a couple of helper methods and more human readable variables.

reneefromhold commented on src/components/presentational/DateRangeCoordinator/index.ts:
```diff
@@ -1 +1 @@
-export { default } from "./DateRangeCoordinator";
\ No newline at end of file
+export { default, DateRangeContext } from "./DateRangeCoordinator";
```
Nice add!

chrisnowak reviewed (approved):
Looks good and the changes make a ton of sense
